soc/nordic/soc.yml

```diff
@@ -37,3 +37,62 @@ family:
           - name: nrf9151
           - name: nrf9160
           - name: nrf9161
+
+# Recovery/erase is only needed once per core. Prevent resetting the cores whilst flashing
+# multiple images until all images for each core have been flashed, this allows security
+# bits to be set during programming without them interfering with additional flashing
+# operations.
+runners:
+  run_once:
+    - '--recover':
+        run: first
+        runners:
+          - nrfjprog
+        groups:
+          - qualifiers:
+              - /(.+)/nrf51([0-9]{3})((.+)|$)/
+          - qualifiers:
+              - /(.+)/nrf52([0-9]{3})((.+)|$)/
+          - qualifiers:
+              - /(.+)/nrf5340/cpunet/
+              - /(.+)/nrf5340/cpuapp/
+              - /(.+)/nrf5340/cpuapp/ns/
+          - qualifiers:
+              - /(.+)/nrf9160/
+              - /(.+)/nrf9160/ns/
+    - '--erase':
+        run: first
+        runners:
+          - nrfjprog
+          - jlink
+        groups:
+          - qualifiers:
+              - /(.+)/nrf51([0-9]{3})((.+)|$)/
+          - qualifiers:
+              - /(.+)/nrf52([0-9]{3})((.+)|$)/
+          - qualifiers:
+              - /(.+)/nrf5340/cpunet/
+          - qualifiers:
+              - /(.+)/nrf5340/cpuapp/
+              - /(.+)/nrf5340/cpuapp/ns/
```

Comment:
qualifier is a board terminology, please refer to SoC or CPU cluster.

Also as mentioned in last review, then we should still not use board information inside soc.yml but limit strictly to SoC and CPU cluster.
That is no `(.+)` to indicate board name.

The use of slash and different meanings within the same context is problematic.
The first and last slash is means something very different than the slashes used inside the match string.

Normally when same character is used for two different purposes, and you need to match a character literally then you would need to escape the character, in this case as: `\/`.

ref: https://github.com/zephyrproject-rtos/zephyr/blob/10f1071802ce8735ac022c4a7cfa46bf1cfaed8f/scripts/west_commands/run_common.py#L263

Reply:
Normally you would but these are stripped by the python code, so it is the same format as snippets. They are a board terminology but this is how they are mapped, it doesn't make any sense to do mapping with a key for soc or cpu cluster, and those cannot apply to the non-secure targets or other common variants

Reply:
`\/` looks ugly and would need to be stripped by the code, I don't see any compelling reason to do that. An alternative character could be used but that would then break snippets and need to be adapted and I also don't see a compelling reason to do that. It can be documented that the name starts and ends with `/` and that these are stripped so do not need escaping somewhere, might already be for snippets

Reply:
Seems snippets actually uses cmake for regex and not python

Reply:
I'm not in favor of `\/` either. I think it only makes sense for something like JavaScript, where regexes are part of the language, so something like this wouldn't fly:
```js
/(.+)/nrf5340/cpuapp/ /* Uncaught SyntaxError: Invalid regular expression flags */
```
For this PR, I don't mind keeping `/.../` as is, but we could consider some alternatives:
* Have different key names for when a value is a regex or not
* Always use regex implicitly

Whatever is chosen here should also be applied for snippets, so it shouldn't be a blocker anyway.

Reply:
why not take the obvious solution.
A regex matching the whole board from start to end would look like:
`^foo/bar$`, `^(.+)/bar$`.
Any literal `^` or `$` would anyway need to be escaped today in order to work correctly.

It would also mean that code like this:
https://github.com/zephyrproject-rtos/zephyr/blob/4372c1b6c0eab652151842fc105d04094bd86224/scripts/west_commands/run_common.py#L264

wouldn't have to explicitly add `^` and `$` around `{match}`

This follows regular regex matching and makes it clear in the yaml that it is actually matching from start to end, and avoid hiding this knowledge in the python code.

Strings without `^` and `$` is then treated as fixed board names.
No reason to invent our custom `/<search-string>/` pattern.

In fact, any command / program which indicates a search using `/` needs to have following `/`s escaped.
Using `sed` to replace `/foo` with `/bar` is `sed 's/\/foo/\/bar/g'`
Searching for `foo/bar` in vim, one must use `/foo\/bar`.

Reply:
> That would cause even more confusion. People would expect those strings to be regexes too.

I don't think more confusing than encapsulating all `/<text>/` pattern as regexes within `^` and `$` inside Python.

But then the only unambiguous solution would be to explicit state when a regex in the list is used, like this:
```
          - qualifiers:
            - name: foo/bar
            - regex: ^/(.+)/nrf5340/cpuapp/ns/$
```

because any other character, like `|`, `?`, etc all have a special meaning in regex.

Reply:
> Always use regex implicitly

Would of course also be an option, and considering that this code only iterates over a very limited set of boards, then I don't think we should actually worry about performance.